numpy/ma/core.py

```diff
@@ -3344,17 +3346,33 @@ def __setitem__(self, indx, value):
             _mask[indx] = mindx
         return
 
-    def __setattr__(self, attr, value):
-        super(MaskedArray, self).__setattr__(attr, value)
-        if attr == 'dtype' and self._mask is not nomask:
-            self._mask = self._mask.view(make_mask_descr(value), ndarray)
-            # Try to reset the shape of the mask (if we don't have a void)
-            # This raises a ValueError if the dtype change won't work
+    # Define so that we can overwrite the setter.
+    @property
+    def dtype(self):
+        return super(MaskedArray, self).dtype
+
+    @dtype.setter
+    def dtype(self, dtype):
+        super(MaskedArray, MaskedArray).dtype.__set__(self, dtype)
+        if self._mask is not nomask:
+            self._mask = self._mask.view(make_mask_descr(dtype), ndarray)
+            # Try to reset the shape of the mask (if we don't have a void).
+            # This raises a ValueError if the dtype change won't work.
             try:
                 self._mask.shape = self.shape
             except (AttributeError, TypeError):
                 pass
 
+    @property
+    def shape(self):
+        return super(MaskedArray, self).shape
+
+    @shape.setter
+    def shape(self, shape):
+        super(MaskedArray, MaskedArray).shape.__set__(self, shape)
```

Comment:
I think this should be just `super(MaskedArray).shape` or perhaps just `ndarray.shape`

Reply:
Edit: I'm wrong. TIL how `super(cls, cls)` works

Reply:
Yes, this annoyance is in part why I wondered if it was a good idea to replace `__setattr__` with this...

Reply:
The advantage is that subclasses are now able to call `Maskedarray.shape.__set__` in the unlikely even they override shape, and it won't accidentally call `ndarray.shape.__set__`

Reply:
Actually, I think this should be `super(MaskedArray, type(self))`